### PR TITLE
Impement change of control mode using WHISTCTL

### PR DESCRIPTION
**Pull request by totto82**
-- Control mode given by WHISTCTL overwrites control mode given in
WCONHIST
-- option for termination of run if changed to BHP control in not
implemented. But ParseContext is used to inform the user and make
throwing optional.

### CODE REVIEW

File: opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp [hunk 1]
```diff
@@ -1184,4 +1184,134 @@ BOOST_AUTO_TEST_CASE(changeBhpLimitInHistoryModeWithWeltarg) {
...
+            " NONE / \n"
+            "WCONHIST\n"
+            " 'P1' 'OPEN' 'ORAT' 5*/ \n"
+            " 'P2' 'OPEN' 'ORAT' 5*/ \n"
+            "/\n"
+            ;
+
+    ParseContext parseContext;
+    DeckPtr deck = parser.parseString(input, parseContext);
+    std::shared_ptr<const EclipseGrid> grid = std::make_shared<const EclipseGrid>(10, 10, 10);
```
Comment (pgdr) [inline opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp]: Should not be a shared pointer.

File: opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp [hunk 2]
```diff
@@ -266,6 +270,21 @@ namespace Opm {
         return treeUpdated;
     }
 
+    void Schedule::handleWHISTCTL(const ParseContext& parseContext, const DeckKeyword& keyword) {
+        for( const auto& record : keyword ) {
+            const std::string& cmodeString = record.getItem("CMODE").getTrimmedString(0);
+            WellProducer::ControlModeEnum controlMode = WellProducer::ControlModeFromString( cmodeString );
+            m_controlModeWHISTCTL = controlMode;
+            const std::string& bhp_terminate = record.getItem("BPH_TERMINATE").getTrimmedString(0);
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: The `getTrimmedSTring()`method returns by value - i.e.
```C++
std::string bhp_terminate = record.getItem("BPH_TERMINATE").getTrimmedString(0);
```

File: opm/parser/eclipse/EclipseState/Schedule/ScheduleEnums.hpp [hunk 3]
```diff
@@ -119,6 +119,7 @@ namespace Opm {
     namespace WellProducer {
 
         enum ControlModeEnum {
+            NONE =     0,
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/ScheduleEnums.hpp]: Maybe a comment here for when `NONE` applies - I guess that is only for `WHISTCTL`.

File: opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp [hunk 4]
```diff
@@ -1184,4 +1184,134 @@ BOOST_AUTO_TEST_CASE(changeBhpLimitInHistoryModeWithWeltarg) {
...
+    BOOST_CHECK_EQUAL(well_p1->getProductionProperties(3).controlMode, Opm::WellProducer::RESV);
+    BOOST_CHECK_EQUAL(well_p2->getProductionProperties(3).controlMode, Opm::WellProducer::RESV);
+
+    // 20 OKT 2008
+    BOOST_CHECK_EQUAL(well_p1->getProductionProperties(4).controlMode, Opm::WellProducer::LRAT);
+    BOOST_CHECK_EQUAL(well_p2->getProductionProperties(4).controlMode, Opm::WellProducer::LRAT);
+
+    // 25 OKT 2008
+    BOOST_CHECK_EQUAL(well_p1->getProductionProperties(5).controlMode, Opm::WellProducer::ORAT);
+    BOOST_CHECK_EQUAL(well_p2->getProductionProperties(5).controlMode, Opm::WellProducer::ORAT);
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp]: Good testing :white_check_mark:

File: opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 2 above)
Comment (atgeirr) [inline opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: (just saw this accidentally, I am not reviewing this)

It is actually ok and not a bug to use a const reference in that context, the returned object will then have the same lifetime as the reference. You may already know this of course, sorry for interrupting!

**Comment by joakim-hove** [comment]
OK; apart from very minor stylistic comments this is ready to be merged - *when* Travis says: :white_check_mark:

**Comment by OPMUSER** [comment]
GRAT?
ThunderBird Signature File

Regards, David Baxendale Email: David.Baxendale@gmail.com

------------------------------------------------------------------------

File: opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp (hunk 1 above)
Comment (totto82) [inline opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp]: I just copied the code from earlier test in the same file. 
Also I thought  Schedule(...) wanted a shared_ptr

File: opm/parser/eclipse/EclipseState/Schedule/ScheduleEnums.hpp (hunk 3 above)
Comment (totto82) [inline opm/parser/eclipse/EclipseState/Schedule/ScheduleEnums.hpp]: Comment added

File: opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp (hunk 2 above)
Comment (totto82) [inline opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp]: The reference is removed

**Comment by joakim-hove** [comment]
Any reason this has stalled?

File: opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp (hunk 1 above)
Comment (totto82) [inline opm/parser/eclipse/EclipseState/Schedule/tests/ScheduleTests.cpp]: After I rebased I understand your comment. It should be fixed now.

**Pull request merged**